### PR TITLE
[XLA:GPU] Add SchedulerMemoryFencing pass to bound LHS peak memory

### PR DESCRIPTION
**Pull request by shawnwang18**
## 1. Current issue: LHS defers weight-gradient work

GPU LHS constructs the schedule bottom-up. When its earlier memory-pressure,
stall, and resource rules do not distinguish the ready candidates, the
`kAsyncDepth` policy prefers the operation whose dependency path contains more
modeled asynchronous latency. This normally helps LHS expose long-latency
collective work for overlap, but in training workload it makes `GW_i` win the
ready-set comparison. Because the schedule is being built in reverse and then
flipped, selecting `GW_i` first places it later in the final execution order.
Repeated decisions can move it past subsequent collectives and produce:

```text
collective-start_i
collective-done_i
collective-start_{i+1}
collective-done_{i+1}
...
GW_i
```

The memory policy does not necessarily prevent this deferral. It participates
only after the locally tracked pressure becomes high, and it may not prefer GW
unless a local choice approaches the configured limit. Each scheduling choice
can therefore appear safe in isolation, even though repeatedly postponing GW
keeps several large buffers live at the same time. The final schedule has no
useful GW/collective overlap and can produce a much larger memory peak or OOM.

## 2. Approach: give LHS a memory-lifetime guardrail

This PR adds `SchedulerMemoryFencing` before LHS and uses the
memory-minimizing schedule as a guide. For each large buffer, it records a
point by which all work that consumes the buffer should have completed, then
prevents LHS from moving any of that work past that point.

The motivation is not to reproduce the reference schedule or prescribe the
exact position of every operation. It is to preserve LHS's freedom to arrange
computation for latency hiding while preventing it from postponing buffer
users so long that memory accumulates. The guardrail is applied to every user
because the buffer cannot be released until its last user executes.

For GW work, LHS remains free to produce the useful overlapping pattern:

```text
collective-start_i
GW_i
collective-done_i
collective-start_{i+1}
```

The guardrail does not force `GW_i` to be inside the start/done interval; it
only prevents GW from being moved arbitrarily far toward the end of the
schedule.

## 3. Benefits

- **Lower memory consumption:** work using a large buffer cannot be postponed
  indefinitely, so the buffer can be released earlier instead of remaining
  live across many later collectives.
- **Better GW/collective overlap:** LHS can place GW computation between an
  asynchronous collective start and done instead of serializing GW after the
  collective operations.
- **Potential performance improvement:** when GPU computation and collective
  communication can execute concurrently, the new placement may hide
  communication latency. The speedup is workload-dependent because of
  critical-path and resource-contention effects.

## 4. Changes in this PR

- Adds the `SchedulerMemoryFencing` pass and runs it immediately before LHS.
- Implements the all-users fence-to-start policy; already implied edges are
  skipped and every new edge points forward in the valid reference schedule.
- Adds the following opt-in debug options:
  - `xla_gpu_experimental_enable_scheduler_memory_fencing=false`
  - `xla_gpu_experimental_scheduler_memory_fencing_threshold_bytes=0`
    (`0` uses 1% of the scheduler memory limit)
  - `xla_gpu_experimental_scheduler_memory_fencing_slack_windows=1`
- Adds unit and GPU scheduling integration tests covering threshold filtering,
  slack 0/1/2, all-users fencing, redundant-edge skipping, schedule validity,
  live-outs, parameters, and multiple GPU configurations.
- Keeps the experimental fence-to-done policy out of this PR; it is isolated in
  draft #45981.

Kind of contribution: 🐛 Bug Fix, ⚡️ Performance Improvement

## 5. Experiment results

The exact validation used Llama3-8B, sequence length 131072, ring CP16 on
4 nodes × 4 GB200 GPUs, per-device batch size 0.5, full rematerialization,
synthetic data, and 20 training steps (`packing=false`, as required by the
current MaxText revision for this CP/synthetic combination). Both rows below
have memory fencing enabled with `slack_windows=1`.

| Workload | Threshold | Added edges | LHS memory (GiB) | Default temp (GiB) | Logical overlap | Mean step (s) | Result |
|---|---:|---:|---:|---:|---:|---:|---:|
| CP16 / 16 GB200 | Auto: 1% (1.260 GiB) | 161 | 95.73 | 81.80 | 130 / 544 relationships | 20.402 | 20 / 20 steps |
| CP16 / 16 GB200 | 500 MB (0.466 GiB) | 608 | 65.08 | 62.98 | 482 / 544 relationships | 19.753 | 20 / 20 steps |
| CP8 / 8 H200 analogue | 500 MB | 608 | 49.29 | — | 32 / 32 mapped GW groups | — | Passed |

Both CP16 runs completed without OOM. The 500 MB threshold captures the
workload's 512 MiB buffer class and reduces default temporary memory by
18.82 GiB relative to the auto threshold.

The logical-overlap counts describe scheduled-HLO ordering inside the preceding
collective start/done interval, not duration-weighted runtime concurrency. The
step-time comparison is directional only: it is a single run on separate
allocations, and no public-HLO speedup is claimed.

Validation: [pipeline](https://gitlab-master.nvidia.com/dl/jet/ci/-/pipelines/58891750),
[auto-threshold job](https://gitlab-master.nvidia.com/dl/jet/ci/-/jobs/368989065),
[500 MB job](https://gitlab-master.nvidia.com/dl/jet/ci/-/jobs/368989068).

## Tests

- `//xla/service:scheduler_memory_fencing_test`
- `//xla/service/gpu:gpu_hlo_schedule_test` for A100, B200, GB200, GB300,
  H100, generic NVIDIA GPU, and RTX 6000 Pro configurations

### CODE REVIEW

File: xla/service/scheduler_memory_fencing.cc [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+namespace xla {
+namespace {
+
+// Per-computation state derived from the reference schedule.
+struct ComputationFencingInfo {
+  // Schedule position of every instruction in the computation.
+  absl::flat_hash_map<const HloInstruction*, int64_t> position;
+  // Async operation windows, ordered by the schedule position of their start
+  // operations.
+  std::vector<HloInstruction*> window_starts;
```
Comment (akuegel) [inline xla/service/scheduler_memory_fencing.cc]: Can you possibly introduce a new struct:

```
struct CollectiveWindow {
  HloInstruction* start;
  int64_t start_position;
  int64_t done_position;
};
```

And use `std::vector<CollectiveWindow> windows;` here?

**Review by akuegel — APPROVED** [review]

File: xla/service/scheduler_memory_fencing.cc (hunk 1 above)
Comment (shawnwang18) [reply xla/service/scheduler_memory_fencing.cc]: Yes，that is better. I introduced AsyncOperationWindow, since this pass also support other async operators.

**Review by dimitar-asenov — APPROVED** [review]

**Comment by akuegel** [comment]
@shawnwang18 can you please rebase?

**Review by akuegel — APPROVED** [review]

File: xla/service/gpu/gpu_hlo_schedule.cc [hunk 2]
```diff
@@ -755,6 +757,19 @@ absl::Status RunLatencyHidingSchedulerPasses(
           ? GpuScheduleCrossesOverlapLimit
           : nullptr);
 
+  const int64_t configured_fencing_threshold_bytes =
+      options.has_xla_gpu_experimental_scheduler_memory_fencing_threshold_bytes()
```
Comment (akuegel) [inline xla/service/gpu/gpu_hlo_schedule.cc:762]: This line is too long, can you please add a comment NOLINTNEXTLINE?

File: xla/debug_options_flags.cc [hunk 3]
```diff
@@ -3142,6 +3146,26 @@ void MakeDebugOptionsFlags(std::vector<tsl::Flag>* flag_list,
       debug_options->xla_gpu_experimental_parallel_async_compute_limit(),
       "This controls how many in-flight asynchronous computations "
       "latency hiding scheduler can schedule."));
+  flag_list->push_back(tsl::Flag(
+      "xla_gpu_experimental_scheduler_memory_fencing_threshold_bytes",
+      int64_setter_for(
+          &DebugOptions::
+              set_xla_gpu_experimental_scheduler_memory_fencing_threshold_bytes),
```
Comment (akuegel) [inline xla/debug_options_flags.cc]: This line is too long, can you please add a comment NOLINTNEXTLINE?

Comment (shawnwang18) [reply xla/debug_options_flags.cc]: done

**Review by akuegel — APPROVED** [review]

**Review by akuegel — APPROVED** [review]
